common/Vectorize.py

```diff
@@ -0,0 +1,307 @@
+                for node_attr in self._NODE_ATTRS:
+                    features.append(str(getattr(graph.nodes[index], node_attr)))
+
+            if include_edge_attrs.lower() == "true" and index in input_edges:
+                for input_edge in input_edges[index]:
+                    for edge_attr in self._EDGE_ATTRS:
+                        features.append(str(getattr(input_edge, edge_attr)))
+
+            concat_feature = " ".join(features)
+
+            networkx_graph.add_node(
+                index, feature = concat_feature)
+
+        # If graph is not connected, introduce a dummy node and connect all 
+        # inputs to it to make it connected.
+        if not networkx.algorithms.is_connected(networkx_graph):
+            max_node_num = max(networkx_graph.nodes())
+
+            networkx_graph.add_node(max_node_num + 1, feature = "Dummy")
+
+            for index in graph.start_node_indices:
+                networkx_graph.add_edge(max_node_num + 1, index)
+
+        return networkx_graph
+
+    def get_graph2vec_embeddings(self, include_edge_attrs = False, 
+                                    include_node_attrs = True):
+        """ Getting embeddings using graph2vec
+
+        Parses models from DB into Graph objects, converts them into networkx
+        graphs and uses graph2vec for getting embeddings for them.
+        Prints the top 20 models closest to every model using cosine similarity.
+
+        Arguments:
+            include_edge_attrs (str) : case insensitive string to denote 
+                whether to include edge attributes of input edge from 
+                _EDGE_ATTRS in feature, if "true" then they are included.
+            include_node_attrs (str) : case insensitive string to denote
+                whether to include node attributes from _NODE_ATTRS in feature,
+                if "true" then they are included.
+
+        Returns:
+            list of Graph objects corresponding to models in the database and
+            a list of embeddings for the same with index correspondence.
+        """
+
+        """ Parameters to Graph2Vec
+
+        WL_ITERATIONS (int) : Depth of rooted sub-graph extracted as feature at
+            every node.
+            Higher values extract more detailed graph structures.
+        ATTRIBUTED (bool) : Boolean to denote whether to use 'feature' 
+            attribute of node for feature building or use 
+            default (degree of node).
+        EPOCHS (int) : Number of epochs to train for.
+        LEARNING_RATE (int) : Learning rate for training.
+        MIN_COUNT (int) : minimum count of feature occurence for it to be 
+            considered in vocabulary.
+        DIMENSIONS (int) : Dimension of the embeddings that are obtained.
+        WORKERS (int) : Number of cores.
+        """      
+
+        WL_ITERATIONS = 3
+        ATTRIBUTED = True
+        EPOCHS = 500
+        LEARNING_RATE = 0.15
+        MIN_COUNT = 5
+        DIMENSIONS = 128
```

Comment:
Do you have a sense for how the convergence/accuracy (for our task of clustering) varies with changes in these values? Also, a comment on how these values were chosen would be helpful (even they are arbitrary for now).

Reply:
Added comment for the same.